core/src/main/scala/kafka/server/KafkaServer.scala

```diff
@@ -577,8 +577,7 @@ class KafkaServer(val config: KafkaConfig, time: Time = Time.SYSTEM, threadNameP
       if (isStartingUp.get)
         throw new IllegalStateException("Kafka server is still starting up, cannot shut down!")
 
-      val canShutdown = isShuttingDown.compareAndSet(false, true)
-      if (canShutdown && shutdownLatch.getCount > 0) {
+      if (shutdownLatch.getCount > 0 && isShuttingDown.compareAndSet(false, true)) {
```

Comment:
Can you please add a comment making it clear that the order of the check is important?